File: samples/speed/speed.rb
$oneMillion = 1000000

class Tester
  attr_accessor :t1, :x

  def foo
    1
  end

  def beginTimer
    @t1 = Time.now.to_f
  end

  def endTimer(s)
    mps = 1.00 / (Time.now.to_f - @t1)
    printf "Ruby %s := %0.2f\n", s, mps
  end

  def testSlot
    beginTimer()
    @x = 1
    ($oneMillion/8).times {
      x; x; x; x;
      x; x; x; x;
    }
    endTimer("slotAccesses       ")
  end

  def testSetSlot
    beginTimer
    @x = 1
    ($oneMillion/8).times {
      @x = 1; @x = 1; @x = 1; @x = 1;
      @x = 1; @x = 1; @x = 1; @x = 1;
    }
    endTimer("slotSets           ")
  end

  def testBlock
    beginTimer
    ($oneMillion/8).times {
      foo; foo; foo; foo;
      foo; foo; foo; foo;
    }
    endTimer("blockActivations   ")
  end

  def testInstantiations
    beginTimer
    ($oneMillion/8).times {
      Tester.new; Tester.new; Tester.new; Tester.new;
      Tester.new; Tester.new; Tester.new; Tester.new;
    }
    endTimer("instantiations     ")
  end

  def testLocals
    beginTimer
    v = 1
    ($oneMillion/8).times {
      v; v; v; v;
      v; v; v; v;
    }
    endTimer("localAccesses      ")
  end

  def testSetLocals
    beginTimer
    v = 1
    ($oneMillion/8).times {
      v = 1; v = 2; v= 3; v= 4;
      v = 1; v = 2; v= 3; v= 4;
    }
    endTimer("localSets          ")
  end

  def test
    puts ""
    testLocals
    testSetLocals
    puts ""
    testSlot
    testSetSlot
    puts ""
    testBlock
    testInstantiations
    printf "Ruby version := \"%s\"", RUBY_VERSION
    puts ""
    puts "// values in millions per second"
    puts ""
  end
end

begin
  Tester.new.test
end

File: samples/speed/speed.py
import sys
import time

oneMillion = 1000000

if sys.version_info >= (3,):
  def xrange(v):
    return range(int(v))

class Tester:
  def foo(self):
    return 1
    
  def beginTimer(self):
    self.t1 = time.time()
    
  def endTimer(self, s):
    mps = 1/(time.time() - self.t1)
    print("Python %s := %0.2f" % (s, mps))
   
  def testSlot(self):
    self.beginTimer()
    self.x = 1
    for i in xrange(oneMillion/8):
      self.x; self.x; self.x; self.x; 
      self.x; self.x; self.x; self.x;
    self.endTimer("slotAccesses       ")

  def testSetSlot(self):
    self.beginTimer()
    self.x = 1
    for i in xrange(oneMillion/8):
      self.x = 1; self.x = 2; self.x = 3; self.x = 4; 
      self.x = 1; self.x = 2; self.x = 3; self.x = 4; 
    self.endTimer("slotSets           ")

  def testBlock(self):
    self.beginTimer()
    for i in xrange(oneMillion/8):
      self.foo(); self.foo(); self.foo(); self.foo(); 
      self.foo(); self.foo(); self.foo(); self.foo()
    self.endTimer("blockActivations   ")

  def testInstantiations(self):
    self.beginTimer()
    for i in xrange(oneMillion/8):
      Tester(); Tester(); Tester(); Tester();
      Tester(); Tester(); Tester(); Tester();
    self.endTimer("instantiations     ")

  def testLocals(self):
    self.beginTimer()
    v = 1
    for i in xrange(oneMillion/8):
      v; v;  v; v;  
      v; v;  v; v;  
    self.endTimer("localAccesses      ")

  def testSetLocals(self):
    self.beginTimer()
    v = 1
    for i in xrange(oneMillion/8):
      v = 1; v = 2; v = 3; v = 4;  
      v = 1; v = 2; v = 3; v = 4;  
    self.endTimer("localSets          ")
    
  def test(self):
    print("")
    self.testLocals()
    self.testSetLocals()
    print("")
    self.testSlot()
    self.testSetSlot()
    print("")
    self.testBlock()
    self.testInstantiations()

    import sys
    print("Python version := \"%i.%i.%i %s %i\"" % tuple(sys.version_info))
    print("")
    print("// values in millions per second")
    print("")


Tester().test()

